pkg/synchromanager/clustersynchro/discovery/discovery_cache.go

```diff
@@ -0,0 +1,446 @@
+			// TODO(iceber): Add limit on fetching server version
+			updated, err := c.fetchServerVersion()
+			if err != nil {
+				utilruntime.HandleError(err)
+			}
+			if !updated {
+				return false
+			}
+
+			updated, _ = c.refetchAllGroupsLocked()
+			return updated
+		}
+	} else {
+		return nil, nil
+	}
+
+	if apiResource, versions := c.getAPIResourceAndVersions(resource); len(versions) != 0 {
+		return &apiResource, versions
+	}
+
+	if refetchFunc != nil && !refetchFunc() {
+		return nil, nil
+	}
+
+	if apiResource, versions := c.getAPIResourceAndVersions(resource); len(versions) != 0 {
+		return &apiResource, versions
+	}
+	return nil, nil
+}
+
+func (c *DynamicDiscoveryManager) GetAllResourcesAsSyncResources() []clusterv1alpha2.ClusterGroupResources {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	sortedResources := make([]schema.GroupResource, 0, len(c.resourceVersions))
+	for gr := range c.resourceVersions {
+		sortedResources = append(sortedResources, gr)
+	}
+	sortGroupResources(sortedResources)
+
+	syncResources := make([]clusterv1alpha2.ClusterGroupResources, 0, len(sortedResources))
+	for _, gr := range sortedResources {
+		resource := clusterv1alpha2.ClusterGroupResources{
+			Group:     gr.Group,
+			Resources: []string{gr.Resource},
+			Versions:  c.resourceVersions[gr],
+		}
+		syncResources = append(syncResources, resource)
+	}
+	return syncResources
+}
+
+func (c *DynamicDiscoveryManager) AttachAllCustomResourcesToSyncResources(resources []clusterv1alpha2.ClusterGroupResources) []clusterv1alpha2.ClusterGroupResources {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	sortedCustomResources := make([]schema.GroupResource, len(c.customResourceGroups))
+	for gr := range c.resourceVersions {
+		if !c.customResourceGroups.Has(gr.Group) {
+			sortedCustomResources = append(sortedCustomResources, gr)
+		}
+	}
+	sortGroupResources(sortedCustomResources)
+
+	syncResources := make([]clusterv1alpha2.ClusterGroupResources, 0, len(resources)+len(sortedCustomResources))
+	for _, resource := range resources {
+		if !c.customResourceGroups.Has(resource.Group) {
+			syncResources = append(syncResources, resource)
+		}
+	}
+
+	for _, gr := range sortedCustomResources {
+		resource := clusterv1alpha2.ClusterGroupResources{
+			Group:     gr.Group,
+			Versions:  c.resourceVersions[gr],
+			Resources: []string{gr.Resource},
+		}
+		syncResources = append(syncResources, resource)
+	}
+	return syncResources
+}
+
+func sortGroupResources(resources []schema.GroupResource) {
+	sort.Slice(resources, func(i, j int) bool {
+		left, right := resources[i], resources[j]
+		if left.Group == right.Group {
+			return left.Resource < right.Resource
+		}
+		return left.Group < right.Group
+	})
+}
+
+func HasListAndWatchVerbs(apiResource metav1.APIResource) bool {
+	var hasList, hasWatch bool
+	for _, verb := range apiResource.Verbs {
+		switch verb {
+		case "list":
+			hasList = true
+		case "watch":
+			hasWatch = true
+		}
+	}
+	return hasList && hasWatch
```

Comment:
vs := sets.NewString(apiResource.Verbs)
return vs.HasAll("list", "watch")

Reply:
updated, This is more concise and clear